servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java

```diff
@@ -137,7 +138,57 @@ protected Publisher() {
      * @see <a href="http://reactivex.io/documentation/operators/filter.html">ReactiveX filter operator.</a>
      */
     public final Publisher<T> filter(Predicate<? super T> predicate) {
-        return new FilterPublisher<>(this, predicate);
+        return filter(() -> predicate);
+    }
+
+    /**
+     * Filters items emitted by this {@link Publisher}.
+     * <p>
+     * This method provides a data transformation in sequential programming similar to:
+     * <pre>{@code
+     *     List<T> results = ...;
+     *     Predicate<K> filter = filterSupplier.get();
+     *     for (T t : resultOfThisPublisher()) {
+     *         if (filter.test(t)) {
+     *             results.add(t);
+     *         }
+     *     }
+     *     return results;
+     * }</pre>
+     *
+     * @param filterSupplier Used to obtain a new {@link Predicate} such that if {@link Predicate#test(Object)} returns
+     * {@code true} the item is emitted to the returned {@link Publisher}. This {@link Predicate} is unique per
+     * subscribe and may therefore keep state and can be used for example to track distinct signals.
+     * @return A {@link Publisher} emits filtered signals observed by this {@link Publisher}.
+     * @see <a href="http://reactivex.io/documentation/operators/filter.html">ReactiveX filter operator.</a>
+     * @see <a href="http://reactivex.io/documentation/operators/distinct.html">ReactiveX distinct operator.</a>
+     */
+    public final Publisher<T> filter(Supplier<? extends Predicate<? super T>> filterSupplier) {
+        return new FilterPublisher<>(this, filterSupplier);
+    }
+
+    /**
+     * Only emits distinct signals observed by this {@link Publisher}.
+     * <p>
+     * {@link Publisher#filter(Supplier)} can be used to customize the distinctness criteria.
```

Comment:
Consider clarifying what "distinct" means, like requirements for equals/hashCode.